llvm/lib/Target/AMDGPU/AMDGPURegBankLegalize.cpp

```diff
@@ -167,6 +168,50 @@ AMDGPURegBankLegalizeCombiner::tryMatchRALFromUnmerge(Register Src) {
   return {nullptr, -1};
 }
 
+// Vgpr0, Vgpr1, ... = G_UNMERGE_VALUES UnmergeSrc (or defined elsewhere)
+// Sgpr0 = G_AMDGPU_READANYLANE Vgpr0
+// Sgpr1 = G_AMDGPU_READANYLANE Vgpr1
+// ...
+// Merge = G_MERGE_VALUES/G_BUILD_VECTOR/G_CONCAT_VECTORS Sgpr0, Sgpr1, ...
+// ->
+// If Vgpr sources are all from same G_UNMERGE_VALUES with sequential indices:
+//   return UnmergeSrc directly
+// Otherwise:
+//   build G_MERGE_VALUES/G_BUILD_VECTOR/G_CONCAT_VECTORS Vgpr0, Vgpr1, ...
+Register AMDGPURegBankLegalizeCombiner::tryFoldRALsIntoMergeLikeInstr(
+    GMergeLikeInstr &Merge) {
+  unsigned NumElts = Merge.getNumSources();
+  SmallVector<Register, 4> VgprSrcs;
+
+  // Collect VGPR sources from READANYLANEs.
+  for (unsigned i = 0; i < NumElts; ++i) {
+    auto [RAL, VgprSrc] =
+        tryMatch(Merge.getSourceReg(i), AMDGPU::G_AMDGPU_READANYLANE);
+    if (!RAL)
+      return {};
```

Comment:
RAL is not used, could use m_GAMDGPUReadAnyLane
```suggestion
    Register RALSrc;
    if (!mi_match(Merge.getSourceReg(i), MRI,
                  m_GAMDGPUReadAnyLane(m_Reg(RALSrc))))
      return {};
```